### PR TITLE
JP-3588: Use Pastasoss datamodel for NIRISS SOSS transform solution

### LINKED ISSUE #8398: Proposed enhancement of extract1d step for NIRISS SOSS data
_Issue [JP-3588](https://jira.stsci.edu/browse/JP-3588) was created on JIRA by [Joseph Filippazzo](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jfilippazzo):_

BACKGROUND: For NIRISS SOSS data, the extract1d step of the pipeline uses (by default) the ATOCA algorithm, which is an optimal spectral extraction routine that disentangles the signals from the overlapping order 1 and order 2 traces. This routine is very sensitive to the wavelength value assigned to each pixel by the ‘wavemap' reference file and the location of the trace centers assigned by the ‘spectrace’ reference file. 
 
PROBLEM: These reference files assume that the pupil wheel for the GR700XD grism is in its nominal position, dispersing the spectra onto the same detector pixels each time. However, the SOSS team’s analysis has shown that SOSS observations exhibit a range of pupil wheel positions (recorded as the PWCPOS value in the FITS header) that do not align with the nominal positions represented in the CRDS reference files.
 
SOLUTION: To account for this misalignment when reducing SOSS data, the SOSS team has developed the PASTASOSS Python package (https://github.com/spacetelescope/pastasoss) that transforms the pixel wavelength values and trace positions to reflect those seen in on-sky observations at a variety of PWCPOS values. We have created a pull request of the jwst pipeline (https://github.com/spacetelescope/jwst/pull/8397) which incorporates this package to solve our problem. It reads in the PWCPOS value from an input JWST DataModel and saves the appropriate ‘wavemap’ and ‘spectrace’ reference files on the fly to perform the optimal extraction. It then updates the locations of the corrected reference files in the DataModel before performing the 1D spectral extraction with the ATOCA algorithm. All input requirements and output data products remain unaffected by this change and the ‘override_wavemap' and ‘override_spectrace’ keyword arguments still function as intended.

### PR DESCRIPTION
**Pull request by tapastro**
Resolves [JP-3588](https://jira.stsci.edu/browse/JP-3588)

Closes #8398 

This PR modifies the algorithm for NIRISS SOSS data transforms, used to align the data with the expected spectral trace positions. Previously, the ATOCA algorithm derived a 3 parameter rotation/offset transform that struggled to deal with certain atypical scenarios; this has been removed, and the https://github.com/spacetelescope/pastasoss team has provided an adapted version of their algorithm for use in the pipeline.

The pastasoss algorithm has introduced a new reference file type, aptly named `pastasoss` holding a `PastasossModel`. These files are not yet on the CRDS OPS server - regression testing to be done using CRDS TEST for now, with the expectation that any okify'ed results will need updating once OPS is mirrored to TEST and/or OPS' CRDS server code is updated to handle the new reftype.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by tapastro** [comment]
Started a run against CRDS-TEST only for `test_niriss_soss.py`: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1717/

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+    'x_min' and 'x_max'. This function is applicable to both individual values
+    and arrays of values. This function will use the min/max values from the
+    training data of the wavecal model.
+    """
+    # scaling the input x values
+    x_scaled = (x - x_min) / (x_max - x_min)
+    return x_scaled
+
+
+def wavecal_model_order1_poly(refmodel, x, pwcpos):
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:54]: add short docstring with parameter definitions, as it wasn't immediately obvious to me what `pwcpos` was at a glance

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+
+def get_soss_traces(
+    refmodel: PastasossModel, pwcpos: float,
+        order: str, subarray: str, interp: bool = True
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    This is the primary method for generate the gr700xd trace position given a
+    pupil wheel positions angle provided in the FITS header under keyword
+    PWCPOS. The traces for a given spectral order are found by perform a
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:263]:
```suggestion
    PWCPOS. The traces for a given spectral order are found by performing a
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+    log.warning("Order not found in reference file trace list.")
+    return -1
+
+
+def get_soss_traces(
+    refmodel: PastasossModel, pwcpos: float,
+        order: str, subarray: str, interp: bool = True
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    This is the primary method for generate the gr700xd trace position given a
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:261]:
```suggestion
    This is the primary method for generating the gr700xd trace position given a
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+    pupil wheel positions angle provided in the FITS header under keyword
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:262]:
```suggestion
    pupil wheel position angle provided in the FITS header under keyword
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+    PWCPOS. The traces for a given spectral order are found by perform a
+    rotation transformation using the refence trace positions at the commanded
+    PWCPOS=245.76 degrees. This methods yield sub-pixel performance and will be
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:265]:
```suggestion
    PWCPOS=245.76 degrees. This method yields sub-pixel performance and will be
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+    improved upon in later interations as more NIRISS/SOSS observations become
+    available.
+
+    Parameters
+    ----------
+    pwcpos : float
+        The pupil wheel positions angle provided in the FITS header under
+        keyword PWCPOS.
+    order : str, optional
+        The spectral order to compute the new traces for. Default is '12'.
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:275]:
```suggestion
        The spectral order for which to compute the new traces. Default is '12'.
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+        Support for order 3 will be added at a later date.
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:276]:
```suggestion
        Order 3 is currently unsupported
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+        # Offset for SUBSTRIP96
+        if subarray == 'SUBSTRIP96':
+            y -= 10
+        # rotated reference trace
+        x_new, y_new = rotate(x, y, pwcpos - refmodel.meta.pwcpos_cmd, origin, interp=interp)
+
+        # wavelength associated to trace at given pwcpos value
+        wavelengths = get_wavelengths(refmodel, x_new, pwcpos, int(order))
+
+        # return x_new, y_new, wavelengths
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:324]:
```suggestion
```

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+    -------
+    Array
+        An array containing the wavelength at each pixel on the detector.
+    """
+    os = np.copy(oversample)
+    xpad = np.copy(padding)
+    ypad = np.copy(padding)
+
+    # No need to compute wavelengths across the entire detector, slightly larger than SUBSTRIP256 will do.
+    dimx, dimy = 2048, 300
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:407]: consider moving hard-coded values into all-caps variable names at top of file

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+    nwave = 5001
+    wave_grid = np.linspace(wavemin, wavemax, nwave)
+
+    # Extrapolate wavelengths for order 1 trace
+    xtrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_x)
+    ytrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_y)
+    spectrace_1 = np.array([xtrace_order1, ytrace_order1, wave_grid])
+
+    # Set cutoff for order 2 where it runs off the detector
+    o2_cutoff = 1783
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:488]: same here

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+    Array, Array
+        The 2D wavemaps and corresponding 1D spectraces
+    """
+    _, order1_x, order1_y, order1_wl = get_soss_traces(refmodel, pwcpos, order='1', subarray=subarray, interp=True)
+    _, order2_x, order2_y, order2_wl = get_soss_traces(refmodel, pwcpos, order='2', subarray=subarray, interp=True)
+
+    # Make wavemap from trace center wavelengths, padding to shape (296, 2088)
+    wavemin = 0.5
+    wavemax = 5.5
+    nwave = 5001
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:479]: same here

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 12]
```diff
@@ -0,0 +1,537 @@
...
+
+    # Extrapolate wavelengths for order 1 trace
+    xtrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_x)
+    ytrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_y)
+    spectrace_1 = np.array([xtrace_order1, ytrace_order1, wave_grid])
+
+    # Set cutoff for order 2 where it runs off the detector
+    o2_cutoff = 1783
+    w_o2_tmp = order2_wl[:o2_cutoff]
+    w_o2 = np.zeros(2040) * np.nan
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/pastasoss.py:490]: where does the 2040 come from?

File: jwst/extract_1d/soss_extract/soss_extract.py [hunk 13]
```diff
@@ -508,24 +518,18 @@ def model_image(scidata_bkg, scierr, scimask, refmask, ref_files, box_weights, s
...
-        A dictionary of the reference file DataModels.
+        A dictionary of the reference file DataModels, along with values for
+        subarray and pwcpos, i.e. the pupil wheel position.
     box_weights : dict
         A dictionary of the weights (for each order) used in the box extraction.
         The weights for each order are 2d arrays with the same size as the detector.
-    subarray : str
-        Subarray on which the data were recorded; one of 'SUBSTRIPT96',
-        'SUBSTRIP256' or 'FULL'.
-    transform : array or list, optional
```
Comment (emolter) [inline jwst/extract_1d/soss_extract/soss_extract.py:518]: Not sure the best place to put this comment, but it looks like `soss_kwargs["transform"]` is no longer used anywhere in the code.  Can `soss_transform` safely be removed from the arguments to `extract_1d_step`?  If so, don't forget to update the docs to reflect that.

File: jwst/extract_1d/soss_extract/pastasoss.py [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+import logging
+import numpy as np
+from scipy.interpolate import interp1d
+from ...datamodels import PastasossModel
+
+log = logging.getLogger(__name__)
+
+
+def get_wavelengths(
+    refmodel: PastasossModel, x: np.ndarray, pwcpos: float, order: int
```
[the remark this answers is not in the document]
Comment (melanieclarke) [reply jwst/extract_1d/soss_extract/pastasoss.py:12]: For what it's worth, I vote for just removing all type hints - I've never found them more helpful than they are annoying to maintain.  But yes, we should file a ticket to sort it out either way; I can file it.

File: jwst/extract_1d/soss_extract/pastasoss.py (hunk 12 above)
Comment (tapastro) [reply jwst/extract_1d/soss_extract/pastasoss.py]: I believe this is the result of using the full frame, minus the 4 pixel reference pixel edges.

File: jwst/extract_1d/soss_extract/soss_extract.py (hunk 13 above)
Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py:518]: I've attempted to fully excise this parameter from both code and the documentation.

File: jwst/extract_1d/soss_extract/pastasoss.py (hunk 9 above)
Comment (tapastro) [reply jwst/extract_1d/soss_extract/pastasoss.py]: Did my best to move all hard-coded values to const vars.

File: jwst/extract_1d/soss_extract/soss_extract.py [hunk 15]
```diff
@@ -38,39 +36,56 @@ def get_ref_file_args(ref_files, transform):
...
-
-    # Warn if not all pixels were corrected
-    msg_warning = 'Some pixels in order {} do not follow the expected dispersion axis'
-    if not flag_o1:
-        log.warning(msg_warning.format(1))
-    if not flag_o2:
-        log.warning(msg_warning.format(2))
+    pastasoss_ref = ref_files['pastasoss']
+    if hasattr(pastasoss_ref.traces[0], "padding"):
+        pad = pastasoss_ref.traces[0].padding
```
[the remark this answers is not in the document]
Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py]: Nice!

File: jwst/extract_1d/soss_extract/pastasoss.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (tapastro) [reply jwst/extract_1d/soss_extract/pastasoss.py]: I agree that type hints without something like mypy/Pydantic for enforcement are not very helpful. They were a carryover and I'm happy to remove them.

**Comment by tapastro** [comment]
Requesting re-reviews after addressing comments!

On the topic of unit tests and documentation: I requested these of the IDT members when they supplied the code, and it never came to fruition. I think it's reasonable to request a paragraph or two of documentation from the NIRISS team, with appropriate links to the pastasoss repo and to the ATOCA paper: https://iopscience.iop.org/article/10.1088/1538-3873/ac8a77. I can file a ticket requesting this.

I have a feeling that NIRISS will not be interested in supplying unit tests, though we can ask.

File: jwst/extract_1d/soss_extract/soss_extract.py [hunk 16]
```diff
@@ -745,13 +743,16 @@ def model_image(scidata_bkg, scierr, scimask, refmask, ref_files, box_weights, s
...
     # Which orders to compute (for modeling, different than extraction).
-    if subarray == 'SUBSTRIP96':
+    if ref_files['subarray'] == 'SUBSTRIP96':
         order_list = [1, 2]
     else:
-        order_list = [1, 2, 3]
+        # Regression in behavior? Used to run on three orders, but
+        # pastasoss does not support order 3
+        # order_list = [1, 2, 3]
+        order_list = [1, 2]
```
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:713]: @tapastro it doesn't look like a change has been made in response to this comment

File: jwst/extract_1d/soss_extract/soss_extract.py [hunk 17]
```diff
@@ -1015,18 +1016,16 @@ def extract_image(decontaminated_data, scierr, scimask, box_weights, bad_pix='mo
     return fluxes, fluxerrs, npixels
```
[the remark this answers is not in the document]
Comment (emolter) [reply jwst/extract_1d/soss_extract/soss_extract.py:971]: @tapastro this one too, unless I missed it

**Review by emolter — COMMENTED** [review]
Still two unresolved comments from my last review.  Otherwise this looks good to me and I'll hit the approve button once those are commented on or otherwise resolved

**Review by melanieclarke — COMMENTED** [review]
It looks like there's a failure in the unit tests - a doctest for `jwst.extract_1d.soss_extract.pastasoss.get_soss_traces`.

I know the team is keen to get this change in this build, so I don't want to hold it up here, but I agree with Ned that this change could use some unit tests and documentation.  I think some more code clean up will probably also be warranted, when the dust settles.  Can we file a ticket for follow up in the next build?

File: jwst/extract_1d/soss_extract/soss_extract.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py]: Hmm, not sure how I missed these. Latest commit should address these.

File: jwst/extract_1d/soss_extract/soss_extract.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (tapastro) [reply jwst/extract_1d/soss_extract/soss_extract.py:951]: ^^

**Comment by tapastro** [comment]
Yes, this bit of code is chock full of technical debt. I can file the ticket.

I had left some machinery in `get_soss_traces` that was left over from its use as a standalone repo, but I realize now that it no longer had any utility, so I've removed some of the offending code. I think it's in a functional state now? 😬

**Review by emolter — APPROVED** [review]

**Review by melanieclarke — APPROVED** [review]
Works for me, since the NIRISS SOSS team is happy with their testing.

**Comment by emolter** [comment]
@tapastro just a note that we should follow up with [JP-3650](https://jira.stsci.edu/browse/JP-3650) once this is merged, perhaps by changing its status to "Ready for Testing" and re-assigning it to someone else?

**Pull request merged**